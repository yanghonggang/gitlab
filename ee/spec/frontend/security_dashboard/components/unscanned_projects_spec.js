import { GlTab } from '@gitlab/ui';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';

import UnscannedProjects from 'ee/security_dashboard/components/unscanned_projects.vue';
import TabContent from 'ee/security_dashboard/components/unscanned_projects_tab_content.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('UnscannedProjects component', () => {
  let wrapper;

  const defaultPropsData = { endpoint: 'foo/bar/endpoint', helpPath: 'foo/bar/help' };

  const defaultState = {
    isLoading: false,
    projects: [],
  };

  const defaultGetters = {
    outdatedProjects: () => [],
    untestedProjects: () => [],
    outdatedProjectsCount: () => 1,
    untestedProjectsCount: () => 1,
  };

  const defaultActions = {
    fetchUnscannedProjects: jest.fn(),
  };

  const factory = ({ getters = {}, propsData = {}, state = {} } = {}) => {
    const store = new Vuex.Store({
      modules: {
        unscannedProjects: {
          namespaced: true,
          actions: defaultActions,
          getters: { ...defaultGetters, ...getters },
          state: { ...defaultState, ...state },
        },
      },
    });

    wrapper = mount(UnscannedProjects, {
      propsData: { ...defaultPropsData, ...propsData },
      store,
      localVue,
      sync: false,
    });
  };

  const outdatedProjectsTab = () => wrapper.find({ ref: 'outdatedProjectsTab' });
  const untestedProjectsTab = () => wrapper.find({ ref: 'untestedProjectsTab' });
  const outdatedProjectsTabContent = () => outdatedProjectsTab().find(TabContent);
  const untestedProjectsTabContent = () => untestedProjectsTab().find(TabContent);
  const outdatedProjectsCount = () => wrapper.find({ ref: 'outdatedProjectsCount' });
  const untestedProjectsCount = () => wrapper.find({ ref: 'untestedProjectsCount' });

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('lifecycle hooks', () => {
    it('fetches projects from the given endpoint when the component is created', () => {
      factory();

      expect(defaultActions.fetchUnscannedProjects).toHaveBeenCalledTimes(1);
      expect(defaultActions.fetchUnscannedProjects.mock.calls[0][1]).toBe(
        defaultPropsData.endpoint,
      );
    });
  });

  describe('header', () => {
    it.each`
      helpPath           | description
      ${'/foo/bar/help'} | ${'not empty'}
      ${null}            | ${'empty'}
    `('matches the snapshot when the "helpPath" prop is $description', ({ helpPath }) => {
      factory({ propsData: { helpPath } });

      expect(wrapper.find('header').element).toMatchSnapshot();
    });
  });

  describe('tab buttons', () => {
    it('shows a tab-button for projects which have outdated security scanning', () => {
      factory();

      expect(outdatedProjectsTab().is(GlTab)).toBe(true);
    });

    it.each`
      type          | projectsCount
      ${'outdated'} | ${outdatedProjectsCount}
      ${'untested'} | ${untestedProjectsCount}
    `(`shows a count of $type projects`, ({ type, projectsCount }) => {
      factory({
        getters: {
          [`${type}ProjectsCount`]: () => 99,
        },
      });

      return wrapper.vm.$nextTick().then(() => {
        expect(projectsCount().text()).toContain(99);
      });
    });

    it('shows a tab-button for projects which have no security scanning configured', () => {
      factory();

      expect(untestedProjectsTab().is(GlTab)).toBe(true);
    });
  });

  describe('tab content', () => {
    beforeEach(factory);

    it.each`
      type          | tabContent
      ${'outdated'} | ${outdatedProjectsTabContent}
      ${'untested'} | ${untestedProjectsTabContent}
    `(
      'passes the "isLoading" prop to the $type projects tab-content component',
      ({ tabContent }) => {
        expect(tabContent().props('isLoading')).toBe(false);

        factory({ state: { isLoading: true } });

        return wrapper.vm.$nextTick(() => {
          expect(tabContent().props('isLoading')).toBe(true);
        });
      },
    );

    it.each`
      type          | tabContent
      ${'outdated'} | ${outdatedProjectsTabContent}
      ${'untested'} | ${untestedProjectsTabContent}
    `(
      'passes the "isEmpty" prop to the $type projects tab-content component',
      ({ type, tabContent }) => {
        expect(tabContent().props('isEmpty')).toBe(false);

        factory({ getters: { [`${type}ProjectsCount`]: () => 0 } });

        return wrapper.vm.$nextTick(() => {
          expect(tabContent().props('isEmpty')).toBe(true);
        });
      },
    );

    it('shows a list of outdated projects', () => {
      factory({
        getters: {
          outdatedProjects: () => [
            {
              description: 'Outdated Projects Group 1',
              projects: [{ fullName: 'Outdated Project One', fullPath: '/outdated-project-1' }],
            },
            {
              description: 'Outdated Projects Group 2',
              projects: [{ fullName: 'Outdated Project Two', fullPath: '/outdated-project-2' }],
            },
          ],
        },
      });

      expect(outdatedProjectsTabContent().element).toMatchSnapshot();
    });

    it('shows a list of untested projects', () => {
      factory({
        getters: {
          untestedProjects: () => [
            { fullName: 'Untested Project One', fullPath: '/untested-1' },
            { fullName: 'Untested Project Two', fullPath: '/untested-2' },
          ],
        },
      });

      expect(untestedProjectsTabContent().element).toMatchSnapshot();
    });
  });
});
