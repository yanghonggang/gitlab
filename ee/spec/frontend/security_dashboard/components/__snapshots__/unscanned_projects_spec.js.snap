// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnscannedProjects component header matches the snapshot when the "helpPath" prop is empty 1`] = `
<header
  class="px-3 pt-3 mb-0"
>
  <h4
    class="my-0"
  >
    
      Project scanning
      
    <!---->
  </h4>
   
  <p
    class="text-secondary mb-0"
  >
    
      Default branch scanning by project
    
  </p>
</header>
`;

exports[`UnscannedProjects component header matches the snapshot when the "helpPath" prop is not empty 1`] = `
<header
  class="px-3 pt-3 mb-0"
>
  <h4
    class="my-0"
  >
    
      Project scanning
      
    <a
      class="gl-link"
      href="/foo/bar/help"
      rel="noopener"
      target="_blank"
      title="Project scanning help page"
    >
      <svg
        aria-hidden="true"
        class="align-top gl-icon s12"
        data-testid="question-icon"
      >
        <use
          href="#question"
        />
      </svg>
    </a>
  </h4>
   
  <p
    class="text-secondary mb-0"
  >
    
      Default branch scanning by project
    
  </p>
</header>
`;

exports[`UnscannedProjects component tab content shows a list of outdated projects 1`] = `
<div
  class="mx-3 my-2 overflow-auto"
  style="height: 15rem;"
>
  <div
    class="mb-3"
  >
    <h5
      class="m-0"
    >
      Outdated Projects Group 1
    </h5>
     
    <ul
      class="list-unstyled mb-0"
    >
      <li
        class="mt-1"
      >
        <a
          class="gl-link"
          href="/outdated-project-1/pipelines"
          rel="noopener"
          target="_blank"
        >
          Outdated Project One
        </a>
      </li>
    </ul>
  </div>
  <div
    class="mb-3"
  >
    <h5
      class="m-0"
    >
      Outdated Projects Group 2
    </h5>
     
    <ul
      class="list-unstyled mb-0"
    >
      <li
        class="mt-1"
      >
        <a
          class="gl-link"
          href="/outdated-project-2/pipelines"
          rel="noopener"
          target="_blank"
        >
          Outdated Project Two
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`UnscannedProjects component tab content shows a list of untested projects 1`] = `
<div
  class="mx-3 my-2 overflow-auto"
  style="height: 15rem;"
>
  <ul
    class="list-unstyled m-0"
  >
    <li
      class="mb-1"
    >
      <a
        class="gl-link"
        href="/untested-1/security/configuration"
        rel="noopener"
        target="_blank"
      >
        Untested Project One
      </a>
    </li>
    <li
      class="mb-1"
    >
      <a
        class="gl-link"
        href="/untested-2/security/configuration"
        rel="noopener"
        target="_blank"
      >
        Untested Project Two
      </a>
    </li>
  </ul>
</div>
`;
